File: Python/TCS34725.py
# Distributed with a free-will license.
# Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
# TCS34725
# This code is designed to work with the TCS34725_I2CS I2C Mini Module available from ControlEverything.com.
# https://www.controleverything.com/products

import smbus
import time

# Get I2C bus
bus = smbus.SMBus(1)

# TCS34725 address, 0x29(41)
# Select Enable register, 0x80(128)
#		0x03(03)	Power ON, RGBC enable, RGBC Interrupt Mask not asserted
#					Wait disable, Sleep After Interrupt not asserted
bus.write_byte_data(0x29, 0x80, 0x03)
# TCS34725 address, 0x29(41)
# Select RGBC Timing register, 0x81(129)
#		0x00(00)	ATIME : 700ms
bus.write_byte_data(0x29, 0x81, 0x00)
# TCS34725 address, 0x29(41)
# Select Wait Time register, 0x83(131)
#		0xFF(255)	WTIME : 2.4ms
bus.write_byte_data(0x29, 0x83, 0xFF)
# TCS34725 address 0x29(41)
# Select Control register, 0x8F(143)
#		0x00(00)	AGAIN is 1x
bus.write_byte_data(0x29, 0x8F, 0x00)

# TCS34725 address 0x29(41)
# Read data back from 0x94(148), 8 bytes
# cData LSB, cData MSB, Red LSB, Red MSB, Green LSB, Green MSB
# Blue LSB, Blue MSB 
data = bus.read_i2c_block_data(0x29, 0x94, 8)

# Convert the data
cData = data[1] * 256 + data[0]
red = data[3] * 256 + data[2]
green = data[5] * 256 + data[4]
blue = data[7] * 256 + data[6]

# Calculate luminance
luminance = (-0.32466 * red) + (1.57837 * green) + (-0.73191 * blue)

# Output data to screen
print "Red Color luminance : %d lux" %red
print "Green Color luminance : %d lux" %green
print "Blue Color luminance : %d lux" %blue
print "IR luminance : %d lux" %cData
print "Ambient Light luminance : %.2f lux" %luminance

File: Java/TCS34725.java
// Distributed with a free-will license.
// Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
// TCS34725
// This code is designed to work with the TCS34725_I2CS I2C Mini Module available from ControlEverything.com.
// https://www.controleverything.com/products

import com.pi4j.io.i2c.I2CBus;
import com.pi4j.io.i2c.I2CDevice;
import com.pi4j.io.i2c.I2CFactory;
import java.io.IOException;

public class TCS34725
{
	public static void main(String args[]) throws Exception
	{
		// Create I2CBus
		I2CBus bus = I2CFactory.getInstance(I2CBus.BUS_1);
		// Get I2C device, TCS34725 I2C address is 0x29(41)
		I2CDevice device = bus.getDevice(0x29);

		// Select enable register
		// Power ON, RGBC enable, wait time disable
		device.write(0x80, (byte)0x03);
		// Select ALS time register
		// Atime = 700 ms
		device.write(0x81, (byte)0x00);
		// Select Wait Time register
		// WTIME : 2.4ms
		device.write(0x83, (byte)0xFF);
		// Select control register
		// AGAIN = 1x
		device.write(0x8F, (byte)0x00);
		Thread.sleep(800);

		// Read 8 bytes of data
		// cData lsb, cData msb, red lsb, red msb, green lsb, green msb, blue lsb, blue msb
		byte[] data = new byte[8];
		device.read(0x94, data, 0, 8);
		// Convert the data
		int cData = ((data[1] & 0xFF) * 256) + (data[0] & 0xFF);
		int red = ((data[3] & 0xFF) * 256) + (data[2] & 0xFF);
		int green = ((data[5] & 0xFF) * 256) + (data[4] & 0xFF); 
		int blue = ((data[7] & 0xFF) * 256) + (data[6] & 0xFF);

		// Calculate final lux
		double luminance = (-0.32466 * red) + (1.57837 * green) + (-0.73191 * blue);

		// Output data to Screen
		System.out.printf("Red Color Luminance   : %d lux %n", red);
		System.out.printf("Green Color Luminance : %d lux %n", green);
		System.out.printf("Blue Color Luminance  : %d lux %n", blue);
		System.out.printf("IR Luminance          : %d lux %n", cData);
		System.out.printf("Ambient Light Luminance : %d lux %n", luminance);
	}
}
